examples/performance/iterator-offsets/iterator-offsets.cxx

```diff
@@ -0,0 +1,291 @@
+        }
+      }
+    }
+    );
+#endif
+
+  // Range based for DataIterator with indices
+  ITERATOR_TEST_BLOCK(
+    "C++11 range-based for (omp)",
+    BOUT_OMP(parallel)
+    for(auto i : result.region(RGN_NOY)){
+      result(i.x,i.y,i.z) = (a(i.x,i.y+1,i.z) - a(i.x,i.y-1,i.z));
+    }
+    );
+
+  // Range based DataIterator
+  ITERATOR_TEST_BLOCK(
+    "C++11 range-based for [i] (omp)",
+    BOUT_OMP(parallel)
+    for(const auto &i : result.region(RGN_NOY)){
+      result[i] = (a[i.yp()] - a[i.ym()]);
+    }
+    );
+
+  ITERATOR_TEST_BLOCK(
+    "C++11 range-based for over Region [i] (omp)",
+    BOUT_OMP(parallel)
+    {
+      for(const auto &i : mesh->getRegion3D("RGN_NOY")){
+        result[i] = (a[i.yp()] - a[i.ym()]);
+      }
+    }
+    );
+
+#ifdef _OPENMP
+  ITERATOR_TEST_BLOCK(
+    "C++11 range-based for [i] with stencil (omp)",
+    BOUT_OMP(parallel)
+    {
+      stencil s;
+      for(const auto &i : result.region(RGN_NOY)){
+        s.mm = nan("");
+        s.m = a[i.ym()];
+        s.c = a[i];
+        s.p = a[i.yp()];
+        s.pp = nan("");
+        result[i] = (s.p - s.m);
+      }
+    }
+    );
+#endif
+  ITERATOR_TEST_BLOCK(
+    "C++11 range-based for [i] with stencil (serial)",
+    stencil s;
+    for(const auto &i : result.region(RGN_NOY)){
+      s.mm = nan("");
+      s.m = a[i.ym()];
+      s.c = a[i];
+      s.p = a[i.yp()];
+      s.pp = nan("");
+      result[i] = (s.p - s.m);
+    }
+    );
+
+  // Region macro
+  ITERATOR_TEST_BLOCK(
+      "Region with stencil (serial)",
+      stencil s;
+      BLOCK_REGION_LOOP_SERIAL(mesh->getRegion3D("RGN_NOY"), i,
+        s.mm = nan("");
+        s.m = a[i.ym()];
+        s.c = a[i];
+        s.p = a[i.yp()];
+        s.pp = nan("");
+
+        result[i] = (s.p - s.m);
+        );
+    );
+#ifdef _OPENMP
+
+  ITERATOR_TEST_BLOCK(
+    "Region with stencil (outside) (parallel section omp)",
+    BOUT_OMP(parallel)
+    {
+      stencil s;
+      BLOCK_REGION_LOOP_PARALLEL_SECTION(
+        mesh->getRegion3D("RGN_NOY"), i,
+        s.mm = nan("");
+        s.m = a[i.ym()];
+        s.c = a[i];
+        s.p = a[i.yp()];
+        s.pp = nan("");
+
+        result[i] = (s.p - s.m);
+        );
+    }
+    );
+
+   ITERATOR_TEST_BLOCK(
+    "Region with stencil (inside) (parallel section omp)",
+    BOUT_OMP(parallel)
+    {
+      stencil s;
+      BLOCK_REGION_LOOP_PARALLEL_SECTION(
+        mesh->getRegion3D("RGN_NOY"), i,
+        s.mm = nan("");
+        s.m = a[i.ym()];
+        s.c = a[i];
+        s.p = a[i.yp()];
+        s.pp = nan("");
+
+        result[i] = (s.p - s.m);
+        );
+    }
+    );
+
+  ITERATOR_TEST_BLOCK(
+    "Region with stencil & raw ints (parallel section omp)",
+    const auto nz = mesh->LocalNz;
+    BOUT_OMP(parallel)
+    {
+      stencil s;
+      BLOCK_REGION_LOOP_PARALLEL_SECTION(
+        mesh->getRegion3D("RGN_NOY"), i,
+        s.mm = nan("");
+        s.m = a[i.ind - (nz)];
```

Comment:
I guess we don't strictly have an `operator[](int)` function so this shouldn't be allowed but works due to implicit conversion?

Reply:
Yes, we might want to mark `Ind3D(int)` as explicit to avoid implicit casting like this